apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java

```diff
@@ -101,6 +101,8 @@ public boolean asChildOf(TraceContext child, Object ignore) {
     private final StringBuilder outgoingHeader = new StringBuilder(TRACE_PARENT_LENGTH);
     private byte flags;
     private boolean discard;
+    @Nullable
+    private ClassLoader applicationClassLoader;
```

Comment:
As we discussed- let's avoid hard reference, so that span leaks are not a big issue if they occur (ie limited to the amount of spans stored in stacks of JVM threads, but not holding entire applications alive).
And as you suggested- in order to avoid reallocating WeakReference, use a WeakConcurrentMap from WeakReference to a WeakReference 🙂